x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java

```diff
@@ -100,4 +109,87 @@ public void testIsAnyLocalAddress() throws Exception {
+
+        validateEndEntityTlsChain(trustStore, certChain, true, true);
+    }
+
+    // Wrap TrustStore in TrustManager[], cast first TrustManager to X509ExtendedTrustManager to call validation methods
+
+    /**
+     * Perform PKIX TLS certificate chain validation. This validates trust and chain correctness, not server hostname verification.
+     * Wrap TrustStore with TrustManager[], and select first element that implements the X509ExtendedTrustManager interface.
+     * Use it to call checkClientTrusted() and/or checkServerTrusted().
+     *
+     * @param trustStore       TrustStore must contain at least one trust anchor (ex: root/sub/cross/link CA certs, self-signed end entity)
+     * @param certChain        Cert chain to be validated using trust anchor(s) in TrustStore. Partial chains are supported.
+     * @param doTlsClientCheck Flag to enable TLS client end entity validation checking.
+     * @param doTlsServerCheck Flag to enable TLS server end entity validation checking.
+     * @throws Exception X509ExtendedTrustManager initialization failed, or cert chain validation failed.
+     */
+    public static void validateEndEntityTlsChain(
+        final KeyStore trustStore, final X509Certificate[] certChain,
+        final boolean doTlsClientCheck, final boolean doTlsServerCheck
+    ) throws Exception {
+        final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance("PKIX", "SunJSSE");
+        trustManagerFactory.init(trustStore);
+        final TrustManager[] trustManagers = trustManagerFactory.getTrustManagers(); // Usually only 1 is returned
+
+        X509ExtendedTrustManager x509ExtendedTrustManager = null;
+        for (final TrustManager trustManager : trustManagers) {
+            if (trustManager instanceof X509ExtendedTrustManager) {
+                x509ExtendedTrustManager = (X509ExtendedTrustManager) trustManagers[0];
+                break; // use the first TrustManager that implements the javax.net.ssl.X509ExtendedTrustManager interface
+            }
+        }
+        if (null == x509ExtendedTrustManager) {
+            throw new UnsupportedOperationException("Expected at least one javax.net.ssl.X509ExtendedTrustManager");
+        }
+
+        // TLS authType is a substring of cipher suite. It controls what OpenJDK EndEntityChecker.java checks in KU/EKU/etc extensions.
+
+        // EKU=null|clientAuth, KU=digitalSignature, authType=DHE_DSS/DHE_RSA/ECDHE_ECDSA/ECDHE_RSA/RSA_EXPORT/UNKNOWN
+        if (doTlsClientCheck) {
+            x509ExtendedTrustManager.checkClientTrusted(certChain, "ECDHE_RSA");
+        }
+
+        // EKU=null|serverAuth, KU=digitalSignature, authType=DHE_DSS/DHE_RSA/ECDHE_ECDSA/ECDHE_RSA/RSA_EXPORT/UNKNOWN
+        // EKU=null|serverAuth, KU=keyEncipherment,  authType=RSA
+        // EKU=null|serverAuth, KU=keyAgreement,     authType=DH_DSS/DH_RSA/ECDH_ECDSA/ECDH_RSA
+        if (doTlsClientCheck) {
```

Comment:
```suggestion
        if (doTlsServerCheck) {
```

Reply:
Good catch. This is a copy and paste mistake on my part. I will fix it.